easier/core/runtime/jit_engine.py

```diff
@@ -0,0 +1,885 @@
+
+    def if_call_method(self, method_name, args, kwargs) -> RuntimeValue:
+        this, *other_args = args
+
+        if not isinstance(this, torch.Tensor):
+            # TODO any cases in FX that non-tensor methods are called?
+            # maybe `a.split().index(3)` -- `tuple.index` is called?
+            raise EasierJitException(
+                "Expect a method of torch.Tensor to be called,"
+                f" but method '{method_name}' of {type(this)} is called"
+            )
+
+        this_method = getattr(this, method_name)
+        # `getattr` on the instance `this` already binds the method to the obj
+        # so we don't pass `this` as an argument anymore.
+        res = this_method(*other_args, **kwargs)
+        return res
+
+    def if_call_module(self, submod, args, kwargs) -> RuntimeValue:
+        """
+        About HaloExchanger:
+        -   input may have been skipped, if this HaloExchanger is recv-only.
+
+        About Reducer:
+        -   input may have been skipped, if this local Reducer is a part of
+            a not-full Reducer
+            -   no elemparts of the local output ElemPart are reduced to.
+            -   no HaloExchanger is inserted for such cases
+
+        -   input may be the output of a HaloExchanger, and is zero-length
+            -   input is not _Skipped
+            -   local Reducer.idx is zero-length, OK for Reducer.forward()
+        """
+        if isinstance(submod, HaloExchanger):
+            # Inserted by EASIER, input always on args[0]
+            arg = args[0]
+            if isinstance(arg, _Skipped):
+                args = (submod.zero_length_input,) + args[1:]
+
+        if isinstance(submod, esr.Reducer):
+            input, opt_out = normalize_reducer_call_into_args(*args, **kwargs)
+            if isinstance(input, _Skipped):
+                # Pre-Reducer HaloExchanger won't be skipped;
+                # Empty input ElemPart won't have reordering Selector.
+                if opt_out is not None:
+                    return opt_out
+
+                else:
+                    dist_env = get_runtime_dist_env()
+                    result_runtime_meta = get_runtime_tensor_metadata(
+                        self.current_node
+                    )
+                    assert isinstance(result_runtime_meta, RuntimeTensorMeta)
+                    assert result_runtime_meta.role == Role.DISTRIBUTED
+
+                    # Same as Reducer.forward()
+                    out = torch.zeros(
+                        result_runtime_meta.shape,
+                        dtype=result_runtime_meta.dtype,
+                        device=dist_env.comm_device
+                    )
+
+                    return out
+
+            # P.S. local Reducer.idx may be zero-length, but forward() and
+            # scatter_reduce_() within can handle it.
+
+        res = submod(*args, **kwargs)
+        return res
+
+
+class FisrtRunNodeEvaluationHandler(NodeEvaluationHandler):
```

Comment:
I suggest we design the handler in a more KISS style, for example, decoupling the node evaluation, metadata verification, and zero batchsize skip, etc.